### DISCUSSION TITLE
respect `completion-regexp-list`

### DISCUSSION THREAD

**Discussion by jdtsmith** [joaotavora/eglot#726] [Ideas]
* Server used:               python-lsp-server
* Emacs version:             GNU Emacs 27.2 
* Operating system:          MacOS
* Eglot version:             20210901.917
* Eglot installation method:       package-install
* Using Doom:                No

The basic completion docs for `try-completion` (also applying to `test-completion`, `all-completions`, etc.) state (emphasis added):

> In addition, to be acceptable, a completion must also match all the
     regular expressions in ‘completion-regexp-list’.  (Unless
     COLLECTION is a function, **in which case that function has to handle
     ‘completion-regexp-list’ itself**.)

The [COLLECTION function](https://github.com/joaotavora/eglot/blob/122dbb9d6ef477c5c5f0351895652b825d6447de/eglot.el#L2357) provided by eglot does not appear to consider `completion-regexp-list`.  This variable is important for completion styles such as [orderless](https://github.com/oantolin/orderless), which put `completion-regexp-list` to heavy work for candidate refinement.  Since most in-built completion tables are obarrays, alists, hashes, etc., the issue really only pertains to programmed completion, such as in eglot. I have a small patch to the `proxies` lambda which implements this, and works well in my testing.  Let me know if you'd like to see a PR.   See also oantolin/orderless#78.  Thanks for eglot!

**Comment by nemethf** [comment]
In Emacs version 26.1, the docstring at the end of try-completion says: Additionally to this predicate, ‘completion-regexp-list’
is used to further constrain the set of candidates.  To me, this means that try-completion internally takes care of filtering candidates according to completion-regexp-list.   However, Eglot doesn't use all-completion.  Isn't that the root cause of this issue?

I'm guessing here, but appropriately changing cl-remove-if-not to all-completion after the "all-completion" comment might just fix the issue.

**Reply by joaotavora** [reply]
That's interesting @nemethf , can you confirm that it does the job.  I need a suitable from-scratch reproduction recipe to be able to experience the incorrect behaviour and witness the correct behaviour after some fix.  The issue-reporting template has intructions for providing such a thing.

**Comment by joaotavora** [comment]
I've converted this to a discussion.  It can be a issue again once all the elements required to make one are provided, as per the template.

**Comment by jdtsmith** [comment]
Thanks for your thoughts.

My apologies for not including a reproduction recipe, this did not seem to me a "BUG" but rather a "FEATURE REQUEST".  Here are the requested elements:

#### Minimal configuration (mandatory)
```sh
 # Type this in a shell to start an Emacs with Eglot configured
 $ emacs -Q -f package-initialize -L /path/to/git-cloned/eglot -l eglot.el 
 ```
 
 ```elisp
;; Eval in *scratch* as directed
;;; Here we demonstrate the role of completion-regexp-list
(eglot-lsp-server- ; hit M-Tab here, notice 5 results 
(setq completion-regexp-list '("-c[hm]")) ; eval this, then re-try above completion: 3 results
(setq completion-regexp-list nil) ; reset to avoid issues
  ```
 
Now open a temporary python file and `M-x eglot` within it.  Try the following:

```python
import os
os.l # Hit M-Tab: 10 results

# Now M-: (setq completion-regexp-list '("d")) And try completion again:
# still 10 results.  This is incorrect behavior

# Now try with M-: (setq completion-regexp-list '("i")) : 4 results starting li

# Don't forget to reset completion-regexp-list!
```

This demonstrates how eglot is _not_ filtering based on `completion-regexp-list` in the same manner as, e.g. emacs-lisp completion.  As mentioned in the documentation for this variable, the capf function is supposed to ensure all results match all of the regexps in this list.   

What I discovered in preparing this example is that eglot does however, seem to be doing _prefix matching_ based on the `completion-regexp-list`.  I was actually surprised about the prefix matching given the documentation advising, for function collection tables, "that function has to handle `completion-regexp-list` itself".  I.e. I didn't expect it to do anything at all, and I'm not sure how or why it does this. But in any case, it's not the correct behavior, as can be seen by comparison with the emacs-lisp example. 

Obviously this is a contrived example that would have little practical purpose.  The real purpose is to support completion styles such as orderless, which use `completion-regexp-list` to filter on multiple filter terms at once.  This does not work with eglot at present. 

That said, before we could proceed with a fix, I think the first order of business is understanding how and where the incorrect prefix filtering is getting triggered.  Any thoughts you had on this would be appreciated.

**Reply by joaotavora** [reply]
Thanks.  I think I can work with these examples.

**Reply by joaotavora** [reply]
If you'd like to make this into an issue again with all the elements you've now provided, you can.  If you prefer to leave this discussion open, that's also fine. Your call.

**Comment by jdtsmith** [comment]
@nemethf I've finally understood your point, thank you!  The documentation regarding collection functions is correct, but elgot's CAPF collection function queries the LSP server and then caches a _normal_ collection list.  It then re-calls try/test-completion with this list.  When called with such a normal (non-function) collection, try/test-completion _do_ respect `completion-regexp-list`.  But since all-completions isn't called as you note, this works only partially.  I suspect this explains the strange partial prefix-style `completion-regexp-list` behavior.  

@joaotavora We can keep it a discussion for now, no worries.

**Comment by jdtsmith** [comment]
I find replacing the `cl-remove-if-not` with a proper `all-completions` call, as suggested by @nemethf, does the trick:

```elisp
          ((eq action t)                                 ; all-completions
	   (all-completions "" (funcall proxies)
			    (lambda (cand)
			      (let* ((item (get-text-property 0 'eglot--lsp-item cand))
				     (text (or (plist-get item :filterText) cand)))
				(and (string-prefix-p probe text completion-ignore-case)
				     (or (not pred) (funcall pred cand)))))))))
```

As a bonus, all-completions is much faster as it's in C.

**Reply by joaotavora** [reply]
Can you provide this as a diff or a patch?

**Reply by jdtsmith** [reply]
```diff
@@ -2363,14 +2363,12 @@ is not active."
           ((null action)                                 ; try-completion
            (try-completion probe (funcall proxies)))
           ((eq action t)                                 ; all-completions
-           (cl-remove-if-not
-            (lambda (proxy)
-              (let* ((item (get-text-property 0 'eglot--lsp-item proxy))
-                     (filterText (plist-get item :filterText)))
-                (and (or (null pred) (funcall pred proxy))
-                     (string-prefix-p
-                      probe (or filterText proxy) completion-ignore-case))))
-            (funcall proxies)))))
+	   (all-completions "" (funcall proxies)
+			    (lambda (cand)
+			      (let* ((item (get-text-property 0 'eglot--lsp-item cand))
+				     (text (or (plist-get item :filterText) cand)))
+				(and (string-prefix-p probe text completion-ignore-case)
+				     (or (not pred) (funcall pred cand)))))))))
        :annotation-function
        (lambda (proxy)
          (eglot--dbind ((CompletionItem) detail kind)
```

BTW, it's also probably useful to pass the `pred` argument of your CAPF collection function on to test- and try-completion, as this may be configured outside of eglot.

**Reply by joaotavora** [reply]
Thanks for the diff.  I think it's a good idea (though I'm very skeptical of the speed boost.  Did you benchmark?)

Can you explain what you mean in your BTW as a diff, too?  Alternatively, clarify what you mean by "this may be...".

**Reply by jdtsmith** [reply]
I mean that at present you ignore the `pred` argument passed into your CAPF `lambda (probe pred action)`.  My diff above overrides but also checks for `pred` and applies it for all-completions.  I'm suggesting  `(try-completion probe (funcall proxies)))` and similar for `test-completion` should have `pred` passed as their final arguments.  

I think @minad is the expert on `all-completions` speed, but as I understand it, it's a reason why vertico and friends are so fast. Probably not a compelling reason on its own, as the round trip via JSON to the LSP server will likely dominate response times.  **Updated thought on speed**: this is true ONLY on the 1st call, since eglot caches results for you.  For something like orderless, where you add more filter terms and the list gets dynamically filtered by regexps as you go, all-completions will certainly be faster than rolling your own with various cl-remove/etc. and lambdas.

**Reply by jdtsmith** [reply]
One more small detail: you can consider whether that last line of my diff should read `(funcall pred text)`, i.e. operate on the `filterText` with the predicate as well.  I don't quite understand when filterText would appear, so I wasn't sure which would be more appropriate.

**Reply by joaotavora** [reply]
```
(defvar joaot/test (make-list 10000 "foo"))

(benchmark-run 50 (length (all-completions "" joaot/test (lambda (f) (equal f "foo")))))
;; (0.7045695599999999 6 0.5299141989999967)  and returns 10000, as expected

(benchmark-run 50 (length (cl-remove-if-not (lambda (f) (equal f "foo")) joaot/test )))
;; (0.581849452 4 0.3667287340000023) and returns 10000, as expected
```

Of course I don't know what you meant by "faster", i.e. what you are comparing too.  I'm also interested in seeing benchmarks that I can actually understand about the speed of "vertico and friends" (what friends?) so that apples can be compared to apples.

That said, I think your all-completions diff is nevertheless good.  I'll make your diff into a patch and commit it under your name, if you don't object. 

As to your "don't ignore pred" suggestion, I've yet to understand what practical effect it has, so until that is clear, I won't touch that part.

**Reply by joaotavora** [reply]
Hmm, taking a closer look at your diff, it seems you changed the body of the lambda. Why?

**Reply by jdtsmith** [reply]
> Hmm, taking a closer look at your diff, it seems you changed the body of the lambda. Why?

Body of the lambda?  Which lambda? The one I pass as the predicate to `all-completions`?

**Reply by jdtsmith** [reply]
> As to your "don't ignore pred" suggestion, I've yet to understand what practical effect it has, so until that is clear, I won't touch that part.

The only practical effect is if some user tool, 3rd party completion style, etc. calls your completion function with its own custom `pred`.   Right now you are ignoring it for test/try-completion, but would be including it for all-completions (with my diff).  

To make my suggestion concrete, I added the proposed changes to my branch (including passing `pred`) and submitted it as PR #733.

**Reply by joaotavora** [reply]
> The only practical effect is if some user tool, 3rd party completion style, etc. calls your completion function with its own custom pred. Right now you are ignoring it for test/try-completion, but would be including it for all-completions (with my diff).

Eglot doesn't support completions styles, and likely never will, because of a fundamental LSP characteristics (for reasons that I can go in to, I've written about this in some issues).  

What other "user tool" do you envision?  

> To make my suggestion concrete, I added the proposed changes to my branch (including passing pred) and submitted it as PR #733.

Thanks, but it was already concrete.  Once we clear the questions, I can make a commit crediting you as the author.

> Body of the lambda? Which lambda? The one I pass as the predicate to all-completions?

Yes.  can you explain the changes to its body?

**Reply by jdtsmith** [reply]
> > Body of the lambda? Which lambda? The one I pass as the predicate to all-completions?
> 
> Yes. can you explain the changes to its body?

This lambda did not exist before.  I wrote it from scratch.  It's job is to:

1. Look for the presence of `filteredText` and substitute that if necessary.
2. Perform the default prefix matching of the `probe` text.
3. Apply any optional predicate that was passed in.

But I now see what you mean: it appears to be doing the same basic job as your former lambda from cl-remove-if-not, just saving one line of code.   If you want to rewrite it that's fine; happy to test a branch.

**Reply by jdtsmith** [reply]
> Eglot doesn't support completions styles, and likely never will, because of a fundamental LSP characteristics (for reasons that I can go in to, I've written about this in some issues).

Sorry I haven't seen what you've written.  But the entire default completion system including CAPF has completion-styles as its final stage.  If you mean eglot won't work well with certain completion styles, that I certainly understand.  I have the same issue in an iPython mode I'm working on.  When prefix completion is the default provided by your backend, you can't magically summon flex or other completions from it. 

The nice thing from my point of view: with the fix discussed here, orderless works great with eglot, to _filter_ the completions which LSP has provided.

**Reply by jdtsmith** [reply]
> What other "user tool" do you envision?

Nothing specific.  Any tool that runs completion at point in eglot, then calls all-completions on the resulting collection with a predicate, as well described in the completion docs.

**Reply by joaotavora** [reply]
> But I now see what you mean: it appears to be doing the same basic job as your former lambda from cl-remove-if-not,

Yes. It's this "appears" that I find slightly worrying. I'd rather not touch the lambda, if you can't demonstrate 100% that it doesn't add or remove existing behavior. If you can, then I'm happy to take it.

> When prefix completion is the default provided by your backend, you can't magically summon flex or other completions from it.

Yes. That's more or less what I meant, in a nutshell.

> The nice thing from my point of view: with the fix discussed here, orderless works great with eglot, to filter the completions which LSP has provided.

I personally would not want that behavior for me: the backend, here, is responsible for choosing the pattern and doing what it wants with it. Adding further filtering may be confusing. The user may be left wondering why his orderless setup returned a completion before, but not now. She might not understand that this particular server is secretly applying its own filtering where, e.g. the orderless pattern is interpreted as a prefix.

That said, I don't oppose this change as I've said, and I don't know how you're compounding completion systems. I take this issue to mean only what its name is: respect `c-r-l` and that seems fine.

> Nothing specific. Any tool that runs completion at point in eglot, then calls all-completions on the resulting collection with a predicate, as well described in the completion docs.

My imagination is too poor. I can't envision this mystery tool or its purpose in practice.

**Reply by jdtsmith** [reply]
> > But I now see what you mean: it appears to be doing the same basic job as your former lambda from cl-remove-if-not,
> 
> Yes. It's this "appears" that I find slightly worrying. I'd rather not touch the lambda, if you can't demonstrate 100% that it doesn't add or remove existing behavior. If you can, then I'm happy to take it.

Since this concerns you I reverted to the original lambda.  It's obvious to me that they do the same thing, but better safe than sorry I suppose.  See PR #733. 

> > The nice thing from my point of view: with the fix discussed here, orderless works great with eglot, to filter the completions which LSP has provided.
> 
> I personally would not want that behavior for me: the backend, here, is responsible for choosing the pattern and doing what it wants with it. Adding further filtering may be confusing. The user may be left wondering why his orderless setup returned a completion before, but not now. She might not understand that this particular server is secretly applying its own filtering where, e.g. the orderless pattern is interpreted as a prefix.

The filtering occurs only as you enter more text, after the completion begins.  Example: in Python when I `import os` then enter `os.` and hit `M-Tab`, I'm not sure what I'm looking for, but I can continue typing random bits, and orderless provides a very fast way to search through and explore the name space.  

> That said, I don't oppose this change as I've said, and I don't know how you're compounding completion systems. I take this issue to mean only what its name is: respect `c-r-l` and that seems fine.

Yep that's the extent of it.

**Reply by joaotavora** [reply]
> The filtering occurs only as you enter more text, after the completion begins. Example: in Python when I import os then enter os. and hit M-Tab, I'm not sure what I'm looking for, but I can continue typing random bits, and orderless provides a very fast way to search through and explore the name space.

I don't know which server or UI you are using, but some servers may only support prefix completion, which means when you type "f" and type tab, they will return "foobarbaz" and "fabrobazo". When you add a "oo" you will see the first, but lose the second.  I've seen this happen.  But I don't know if orderless or whatever toys you're using cache things and then search from there and doesn't re-request.  They shouldn't because the server may be truncating or using some mystery heuristic).  Anyway if you're happy and it doesn't happen to you in your particular setup/server, that's great.

**Reply by joaotavora** [reply]
> Since this concerns you I reverted to the original lambda. It's obvious to me that they do the same thing, but better safe than sorry I suppose. See PR #733.

I see now, since you changed variable names it was hard to see. I prefer not to mix cosmetic and functional goals.  This would have dirtied up the code archeology. Also, there's a subtle behavior difference: the order of the conditions in the `and` is reversed, which have speed differences or even funcional ones if `pred` happens to have side-effects.

So I've pushed a commit that closes this and credited you and @nemethf  in the commit message.

**Reply by jdtsmith** [reply]
> > The filtering occurs only as you enter more text, after the completion begins. Example: in Python when I import os then enter os. and hit M-Tab, I'm not sure what I'm looking for, but I can continue typing random bits, and orderless provides a very fast way to search through and explore the name space.
> 
> I don't know which server or UI you are using, but some servers may only support prefix completion, which means when you type "f" and type tab, they will return "foobarbaz" and "fabrobazo". When you add a "oo" you will see the first, but lose the second. I've seen this happen. But I don't know if orderless or whatever toys you're using cache things and then search from there and doesn't re-request. They shouldn't because the server may be truncating or using some mystery heuristic). Anyway if you're happy and it doesn't happen to you in your particular setup/server, that's great.

In fact it is elgot that does the caching!  The LSP server I'm using (pylsp) is indeed doing simple prefix completion.  But once the completion UI pops up (either automatically, or on demand), the same CAPF table lambda eglot initially created [1] is queried over and over again to continue refining the search (in the case of orderless, by setting `completion-regexp-list` and sending lots of  `all-completions` requests).   

It's surprisingly straightforward in practice: prefix matching to "get started" then refine/down-select in a flexible manner.  It's why I'm not at all averse to hitting `M-Tab` when I know the results will be a rather large list (e.g. after `os.`): I can quickly narrow it down using flex/regexp/initials/etc. matching client side even if I only sort of know what I'm looking for.  

If the server were doing some fancy/flex style matching, I guess I'd have the same mental model: "Hey LSP server, whatever you do before my UI pops up is up to you.  After that, my client and I will cache the results and take over the filtering.  Thanks."

[1] I once asked Stefan Monnier how long a completion at point collection function lives.  His answer: as long as it is needed.

**Reply by joaotavora** [reply]
> In fact it is elgot that does the caching!

That's true, but I don't think it keeps the cache as you add more characters.  In my example, I think (haven't verified) than when you add the `oo`, you get another request to the server, _not_ some kind of filtering of what it had already sent for `foo`.

If your mental model works for you, that's great.  For me it's a very confusing model, but that's just me.  I prefer to think "oh, this server doesn't support flex matching, pity.  I'll find another one that does".

**Reply by jdtsmith** [reply]
> > In fact it is elgot that does the caching!
> 
> That's true, but I don't think it keeps the cache as you add more characters. 

Corfu does, as I just verified by printing a message on cache retrieval.   I also tried with company — it keeps hitting the LSP server on each new character. 

> In my example, I think (haven't verified) than when you add the `oo`, you get another request to the server, _not_ some kind of filtering of what it had already sent for `foo`.

It all depends on the completion UI, whether it keeps your collection table around during completion, or dumps it and asks for a new one.  I definitely prefer the (much faster) "keep it around" behavior.

**Reply by jdtsmith** [reply]
On this topic, there is one drawback for completion UI's which keep the collection table around for the duration.  The current eglot cache never gets invalidated, and becomes incorrect if the original text which started the completion is deleted or altered.  This doesn't show up with company, because company spins off a new completion table function on each new character event.   But it leads to some surprising results with other completion UI's.  See minad/corfu#59.  

I tried a simple cache invalidation strategy:

```elisp
      (lambda (probe pred action)
	 (unless (string-empty-p probe)
	   (if last-probe
	       (unless (string-prefix-p last-probe probe)
		 (setq cached-proxies :none
		       last-probe probe))
	     (setq last-probe probe)))
...
```

and it worked well.  This code would not affect company, because it dumps the entire collection table every time more text is entered (a heavy-handed way to invalidate the cache).  But it would address the confusing behavior shown in the issue mentioned. Thoughts?

**Reply by joaotavora** [reply]
@jdtsmith that perhaps only one of the reasons I don't cache completions (If I understand it completely). 

The main one is that with LSP one cannot probably make the assumption that adding a new character will _not_ bring more completions (the current completion list may be incomplete, and one could easily envision a hypothetical server-side regexp completion style where adding the character '?' changes the meaning of things completely). Or even a completely different set of completions, for whatever reason, including server bugs. According to LSP, that is up to the LSP server last time I checked (it may have changed). 

So whenever the pattern changes, Eglot/Emacs must go back to the source to get more completions.  I really don't see a way around this unless Eglot/Emacs has full information to start with, because one can only filter reliably if starting with the full information.  But in LSP, the client _doesn't_ have this full information -- that's the whole point of LSP.  

Implementing fancy client-specified completion styles -- like `flex`, `orderless`, `toms-regexps`, whatever -- that work reliably for all servers requires overhauling the LSP protocol.  For example, if the protocol specified it and LSP servers could respect something akin to `completion-regexp-list`, Emacs could just hand it to them and they would do the right thing and always return the completions we want, filtered the way you want them.

**Reply by joaotavora** [reply]
Having a cursory look at the Gorfu issue you opened, it seems you bumped exactly into the problem I described.

**Reply by joaotavora** [reply]
Here's another piece of information you may be missing.  The LSP server will only return new completions if the text is actually in the buffer (in the "shared" LSP document between client and server).  So tweaking the pattern  in a widget without actually changing the buffer will not request new completions: the change has to be sent over to the server, processed by it, and then a new `textDocument/completion` request is issued that hopefully takes the new LSP document in consideration.  If you "cache" any of these resulting tables, you might get lucky or you might bump into trouble (as I've explained already).

**Reply by joaotavora** [reply]
@jdtsmith please make a PR of the code change you propose, it's easier to analyse there.

**Reply by jdtsmith** [reply]
Thanks for the thoughts.  Eglot does cache LSP completions, for as long as its collection lambda is alive.  With company, that's not very long (several calls to test/try/all-completion).  For other UIs like corfu and completion in minibuffer, it's much longer.  I think @minad makes a good argument that correct caching is up to the backend.  Apparently VSCode caches (for speed).  

I see your point about the client being unable to predict whether LSP will return the same completion set.  I hadn't thought about incomplete lists of candidates (based on an LSP limit).  Right now company is "doing this for" eglot.

Let me make a PR and perhaps a few of us can discuss there.

**Reply by joaotavora** [reply]
> makes a good argument that correct caching is up to the backend. Apparently VSCode caches (for speed).

I don't know that argument: I just know that `completion-at-point-functions` 's docstring says that the return value of the functions you put there is to complete an "entity":
``` 
....
or a list of the form (START END COLLECTION . PROPS), where:
 START and END delimit the entity to complete and should include point,
 COLLECTION is the completion table to use to complete the entity, and
 PROPS is a property list for additional information.
....
```

It was around this docstring, not someone's opinions, that I built the Eglot table (and I've also built completion UIs btw).  COLLECTION is the completion entity to complete *the entity*, not _the entity and some variations on that entity_.

So it you change the "entity" to complete and don't re-request `completion-at-point-functions` you'll run into problems.  

Is experimenting with adding a few characters "changing the entity"? Sometimes, maybe not.  Company seems to think it is, and VSCode followers would think it's not? 

 What about removing characters from it? I think it is, undeniably.  It's pretty clear that the UI has to re-request a table to me.  The entity to complete has changed.

**Comment by jdtsmith** [comment]
The speedup comes primarily from searching through a number of regexps, which all-completions does internally.  Orderless benefits the most from this, as it reruns this type of search over and over on every character input.  And since it compiles regexps from user input with a custom dispatcher, they do tend to get some length and complexity.  The general vertico/embark/consult/orderless group of packages all use native completion internally for speed.  

Here's an example that shows a bit more than 3x speedup for me:

```elisp
(let ((candidates (cl-loop
		  repeat 10000
		  for len = (+ 5 (random 12))
		  for word = (cl-loop repeat len concat (string (+ ?a (random 26))))
		  collect word))
      (completion-regexp-list '("[a-d]$"
				"r..[pqr]"
				"\\([^mpqr]\\)\\{3\\}.*[b-n]\\1[a-x]"
				"[a-k][k-z]\\{5\\}.*?[a-m]")))

  (list
   (benchmark-run 100 (length (all-completions "" candidates)))
   (benchmark-run 100 (length
    (cl-remove-if-not
     (lambda (cand)
       (cl-every (lambda (regexp)
		   (string-match-p regexp cand))
		 completion-regexp-list))
     candidates)))))
```

`((0.339078 0 0.0) (1.175021 0 0.0))`

**Reply by joaotavora** [reply]
> The speedup comes primarily from searching through a number of regexps,

Right: it's faster when there are regexps, but slower when there are none.  This is why I asked what your benchmark case is.

**Reply by jdtsmith** [reply]
Indeed, the very case that motivated this discussion!  With orderless, there are _always_ regexps.  Sometimes lots.  I recognize that other completion environments don't have this particular issue.

**Reply by jdtsmith** [reply]
__

> > The speedup comes primarily from searching through a number of regexps,
> 
> Right: it's faster when there are regexps, but slower when there are none. This is why I asked what your benchmark case is.

Just an FYI, on my system all-completions is about 10% faster on your benchmark.  So similar speed.  How does it compare for you with the regexps?  It's clearly quite dependent on how selective the early regexps are in my testing.

**Reply by joaotavora** [reply]
When regexps are involved, all-completions is much faster, as I've previously reported, I think.

**Comment by joaotavora** [comment]
> The general vertico/embark/consult/orderless group of packages all use native completion internally for speed.

All completion tools I know use "native" completion (which I think is what you mean by calling into C code) to a certain degree. Here you are grouping completion UIs and completion tables and completion styles, if I'm not mistaken.  

I understand that your general perception is that this combo is "fast", but as I said, I would like to see objective benchmarks and numbers, preferably measuring each of these in isolated and controlled conditions.  But not need to provide this to for this particular matter regarding `completion-regexp-list`, it's just a general with of mine :-)

**Reply by jdtsmith** [reply]
The comparison is not against other completion tools, but other _selection_ tools for buffers/files/variables/imenu listings/bookmarks/etc.  Things like consult for example use completion to make filtering through these as fast as completion by using the speed of _all-completions_.  I don't have specific benchmarks.

**Reply by joaotavora** [reply]
> The comparison is not against other completion tools, but other selection tools for buffers/files/variables/imenu listings/bookmarks/etc. Things like consult for example use completion to make filtering through these as fast as completion by using the speed of all-completions. I don't have specific benchmarks.

In my limited understanding consult is a sort of completion table and vertico is an UI, both can be used with any other table/UI. So I'm still very confused (but that's ok).

When talking about speed, some kind of quantitative really is generally  useful to understand what one is talking about, else my dentist is faster because she uses elephants :)

**Comment by jdtsmith** [comment]
Again without benchmarks, the quantitive (but broad) category is: how to rapidly select from a large number of possibilities by prefix, sub-string, multiple-regexps, etc. Those possibilities could be "things to insert in the buffer" aka completion, but also many many others.  I think my example above proves that at least for some limited aspects of this problem space, using the internal completion (written in C) is a faster solution than reasonably efficient Elisp.  And I really don't find that surprising at all.  But maybe your dentist's elephants can make some more benchmarks for her to investigate :).  This is for Emacs 27; I haven't tried the new JIT compiled Emacs 28, which may reduce the distance between approaches.  

Consult is basically a tool that takes all kinds of Emacs selection problems, and runs them through mini-buffer completion (with all its configurability).  Files (like ido/fido/etc.).  Buffers.  Macros to run.  Bookmarks.  Matching lines.  Imenu items.  Markers.  Command history.  (Rip)Grep search results.  Yank history.  Registers.  You get the idea. 

Once I get it working well maybe I'll make a small video of eglot + corfu[1] + orderless doing their thing.  

[1]  corfu is the vertico sibling which provides  in-buffer completion, kind of like company

**Comment by jdtsmith** [comment]
For posterity, this got fixed in PR #733 and 9b63f21e75dc8222b5dfc4006ce8559bb73532cb.
